Userland/Applications/SystemMonitor/ProcessModel.cpp

```diff
@@ -469,54 +469,104 @@ void ProcessModel::update()
+                    state.time_user = thread.time_user;
+                    state.time_kernel = thread.time_kernel;
+                    state.kernel = process.kernel;
+                    state.executable = process.executable;
+                    state.name = thread.name;
+                    state.command = read_command_line(process.pid);
+                    state.uid = process.uid;
+                    state.state = thread.state;
+                    state.user = process.username;
+                    state.pledge = process.pledge;
+                    state.veil = process.veil;
+                    state.cpu = thread.cpu;
+                    state.priority = thread.priority;
+                    state.amount_virtual = process.amount_virtual;
+                    state.amount_resident = process.amount_resident;
+                    state.amount_dirty_private = process.amount_dirty_private;
+                    state.amount_clean_inode = process.amount_clean_inode;
+                    state.amount_purgeable_volatile = process.amount_purgeable_volatile;
+                    state.amount_purgeable_nonvolatile = process.amount_purgeable_nonvolatile;
+                    state.syscall_count = thread.syscall_count;
+                    state.inode_faults = thread.inode_faults;
+                    state.zero_faults = thread.zero_faults;
+                    state.cow_faults = thread.cow_faults;
+                    state.unix_socket_read_bytes = thread.unix_socket_read_bytes;
+                    state.unix_socket_write_bytes = thread.unix_socket_write_bytes;
+                    state.ipv4_socket_read_bytes = thread.ipv4_socket_read_bytes;
+                    state.ipv4_socket_write_bytes = thread.ipv4_socket_write_bytes;
+                    state.file_read_bytes = thread.file_read_bytes;
+                    state.file_write_bytes = thread.file_write_bytes;
+                    state.cpu_percent = 0;
+
+                    auto thread_data = m_threads.ensure(thread.tid, [&] { return make_ref_counted<Thread>(**process_state); });
+                    thread_data->previous_state = move(thread_data->current_state);
+                    thread_data->current_state = move(state);
+                    if (auto maybe_thread_index = (*process_state)->threads.find_first_index(thread_data); maybe_thread_index.has_value()) {
+                        (*process_state)->threads[maybe_thread_index.value()] = thread_data;
+                    } else {
+                        (*process_state)->threads.append(thread_data);
+                    }
+
+                    live_tids.set(thread.tid);
+                }
+            } else {
+                // NOTE: If there are no threads left in a process this is an indication
+                // for a zombie process, so it should be handled differently.
                 ThreadState state(**process_state);
-                state.tid = thread.tid;
+                state.tid = process.pid;
```

Comment:
It feels wrong to report a thread ID if there isn't one. Can we set it to 0 instead?

Reply:
I tested this and it seems to be the actually another source of a crash we have with zombie processes - TID 0 is invalid and TID that equals the PID actually makes sense to me because the first thread of each process has a TID which is the same for the PID, so we essentially simulate a process with one thread although it does not have any.

Reply:
> it seems to be the actually another source of a crash we have with zombie processes - TID 0 is invalid

Is that crash fixable? It probably stems from the fact that we're trying to shoehorn a process inside of a thread structure...

> TID 0 is invalid and TID that equals the PID actually makes sense to me because the first thread of each process has a TID which is the same for the PID

Sure, but that thread doesn't exist anymore, so TID=PID actually no longer makes sense to me at that point.

Reply:
I tried for a couple of hours and the answer is sure, yes, this is fixable, but not with an easy one-line solution. I still need to learn more about the whole LibGUI model abstractions as this is completely new to me.

And yes, the crash is happening because the whole process model and its entries are tied to threads as the main representative rather using processes, just like you said.

Can we put a FIXME note on this and work this out later on?

Reply:
Yes, FIXME sounds fine!